rim/static/rim/js/dynamic_fields.js

```diff
@@ -0,0 +1,99 @@
+
+    for (const d of json_fields['displays'])
+        document.getElementById('display_inputs').innerHTML += display_form(displays++, d["HeightInMillimeters"],
+            d["ManufacturerName"], d["ProductCodeID"], d["Resolution"], d["SerialNumberID"], d["UserFriendlyName"],
+            d["WidthInMillimeters"]);
+}
+
+function addInputs(field) {
+    switch (field) {
+        case 'storage': document.getElementById('storage_inputs').innerHTML += storage_form(storages++, '', '');
+            break;
+        case 'gpu': document.getElementById('gpu_inputs').innerHTML += gpu_form(gpus++, '', '');
+            break;
+        case 'nic': document.getElementById('nic_inputs').innerHTML += nic_form(nics++, '', '', '');
+            break;
+        case 'display': document.getElementById('display_inputs').innerHTML += display_form(displays++, '', '', '', '', '', '', '');
+            break;
+
+    }
+}
+
+function submitAddEditForm() {
+    let storage_json = [], gpu_json = [], nic_json = [], display_json = [];
+
+    for (let i = 0; i < storages; i++)
+        if (document.getElementById('storage_model_' + i).value)
+            storage_json.push({
+                "Model": document.getElementById('storage_model_' + i).value,
+                "Size": document.getElementById('storage_size_' + i).value
+            });
+
+
+    for (let i = 0; i < gpus; i++)
+        if (document.getElementById('gpu_name_' + i).value)
+            gpu_json.push({ "Name": document.getElementById('gpu_name_' + i).value });
+
+    for (let i = 0; i < nics; i++)
+        if (document.getElementById('nic_mac_' + i).value)
+            nic_json.push({
+                "MAC": document.getElementById('nic_mac_' + i).value,
+                "Name": document.getElementById('nic_name_' + i).value,
+                "Type": document.getElementById('nic_type_' + i).value
+            })
+
+    for (let i = 0; i < displays; i++)
+        if (document.getElementById('display_serial_' + i).value)
+            display_json.push({
+                "HeightInMillimeters": document.getElementById('display_height_' + i).value,
+                "ManufacturerName": document.getElementById('display_manufacturer_' + i).value,
+                "ProductCodeID": document.getElementById('display_code_' + i).value,
+                "Resolution": document.getElementById('display_resolution_' + i).value,
+                "SerialNumberID": document.getElementById('display_serial_' + i).value,
+                "UserFriendlyName": document.getElementById('display_name_' + i).value,
+                "WidthInMillimeters": document.getElementById('display_width_' + i).value
+            })
+
+    dynamic_json = { "storage": storage_json, "GPU": gpu_json, "network_cards": nic_json, "displays": display_json }
+    document.getElementById('id_json_fields').value = JSON.stringify(dynamic_json);
+    document.add_edit_form.submit();
+}
+
+setTimeout(() => populateInputs(), 500);
```

Comment:
It is less than ideal to use a timeout to wait for the DOM to be ready, since we don't know in general how long that might take. It may take 1ms or 1 minute, and we'd like our code to work in either case.

Consider `$(document).ready(populateInputs);` instead. This way, we can call the function once the DOM is reported to be ready, however long that might take.